// zudoku/lib/ui/Secret.js
import { j as e } from "../jsx-runtime-DywqP_6a.js";
import { EyeOffIcon as u, EyeIcon as b, CheckIcon as y, CopyIcon as k } from "lucide-react";
import { useState as v } from "react";
import { c as s } from "../cn-dYga0KKN.js";
import { u as z } from "../useCopyToClipboard-B_085nfO.js";
import { Button as c } from "./Button.js";
const N = ({
  secret: o,
  revealed: l = !1,
  previewChars: n = 5,
  className: i
}) => /* @__PURE__ */ e.jsxs("span", { className: s("w-full truncate", i), children: [
  /* @__PURE__ */ e.jsx(
    "div",
    {
      className: s(
        "w-40 inline-block md:w-fit",
        l ? "" : "opacity-50"
      ),
      children: l ? o.slice(0, n === 0 ? o.length : -n) : "•••• ".repeat(
        o.slice(
          0,
          n === 0 ? o.length : -n
        ).length / 5
      ) + "•".repeat(
        o.slice(
          0,
          n === 0 ? o.length : -n
        ).length % 5
      )
    }
  ),
  n > 0 && /* @__PURE__ */ e.jsx("span", { children: o.slice(-n) })
] }), B = ({
  secret: o,
  preview: l = 5,
  className: n,
  status: i,
  onCopy: r,
  revealed: a,
  onReveal: d,
  noReveal: m
}) => {
  const x = Math.abs(l), [p, f] = v(!1), [j, g] = z(), t = a ?? p;
  return /* @__PURE__ */ e.jsxs(
    "div",
    {
      className: s(
        "flex gap-2 items-center text-sm border rounded-md px-1",
        n
      ),
      children: [
        /* @__PURE__ */ e.jsxs("div", { className: "font-mono w-full h-9 items-center flex px-2 text-xs gap-2", children: [
          i && /* @__PURE__ */ e.jsx(
            "div",
            {
              className: s(
                "rounded-full shrink-0 w-2 h-2 mr-2",
                i === "active" && "bg-emerald-400",
                i === "expired" && "bg-neutral-200",
                i === "expiring" && "bg-yellow-400",
                i === "revoked" && "bg-red-400",
                i === "none" && "opacity-0"
              )
            }
          ),
          /* @__PURE__ */ e.jsx(
            N,
            {
              secret: o,
              revealed: t,
              previewChars: x
            }
          )
        ] }),
        m !== !0 && /* @__PURE__ */ e.jsx(
          c,
          {
            variant: "ghost",
            onClick: () => {
              f((h) => !h), d?.(!t);
            },
            size: "icon",
            children: t ? /* @__PURE__ */ e.jsx(u, { size: 16 }) : /* @__PURE__ */ e.jsx(b, { size: 16 })
          }
        ),
        /* @__PURE__ */ e.jsx(
          c,
          {
            variant: "ghost",
            onClick: () => {
              g(o), r?.(o);
            },
            size: "icon",
            children: j ? /* @__PURE__ */ e.jsx(y, { size: 16 }) : /* @__PURE__ */ e.jsx(k, { size: 16 })
          }
        )
      ]
    }
  );
};
export {
  B as Secret,
  N as SecretText
};
