// zudoku/lib/MdxPage-Ud1w_GV-.js
import { j as e } from "./jsx-runtime-DywqP_6a.js";
import { s as T } from "./index-DhL3LhWV.js";
import { H as M } from "./index.esm-BnYHxCYC.js";
import { EditIcon as P } from "lucide-react";
import { useEffect as S } from "react";
import { B as L } from "./Button-DTDRi4CT.js";
import { C as k } from "./CategoryHeading-BGCNu3VK.js";
import { T as C, H as h } from "./ErrorAlert-xpVkFS_n.js";
import { T as H } from "./Toc-Cf5UojVb.js";
import { u as D, a as I } from "./ZudokuContext-BcqSiygJ.js";
import { P as R } from "./Pagination-Cdi_qiXc.js";
import { u as A } from "./useExposedProps-B8NHdNRq.js";
const _ = {
  h2: ({ children: t, id: s }) => /* @__PURE__ */ e.jsx(h, { level: 2, id: s, registerNavigationAnchor: !0, children: t }),
  h3: ({ children: t, id: s }) => /* @__PURE__ */ e.jsx(h, { level: 3, id: s, registerNavigationAnchor: !0, children: t })
}, Y = ({
  mdxComponent: t,
  frontmatter: s = {},
  defaultOptions: d,
  __filepath: a,
  tableOfContents: n,
  excerpt: j
}) => {
  const v = D()?.categoryLabel, o = s.title, p = s.description ?? j, c = s.category ?? v, w = s.toc === !1 || d?.toc === !1, b = o ?? n.find((i) => i.depth === 1)?.value, N = s.disable_pager ?? s.disablePager ?? d?.disablePager ?? !1, f = s.showLastModified ?? d?.showLastModified ?? !0, r = s.lastModifiedTime ? new Date(s.lastModifiedTime) : null, l = s.suggestEdit !== !1 && (s.suggestEdit ?? d?.suggestEdit), g = l ? l.url.replaceAll("{filePath}", a) : null, y = l ? l.text || "Edit this page" : null, u = n.find((i) => i.depth === 1)?.children ?? // if `title` is provided by frontmatter it does not appear in the table of contents
  n.filter((i) => i.depth === 2), E = !w && u.length > 0, { prev: m, next: x } = I();
  return S(() => {
    if (process.env.NODE_ENV === "development")
      return window.__getReactRefreshIgnoredExports = ({ id: i }) => {
        if (i.endsWith(a))
          return ["frontmatter", "tableOfContents"];
      }, () => {
        window.__getReactRefreshIgnoredExports = void 0;
      };
  }, [a]), /* @__PURE__ */ e.jsxs(
    "div",
    {
      className: "grid grid-cols-1 xl:grid-cols-(--sidecar-grid-cols) gap-8 justify-between",
      "data-pagefind-filter": "section:markdown",
      "data-pagefind-meta": "section:markdown",
      children: [
        /* @__PURE__ */ e.jsxs(M, { children: [
          /* @__PURE__ */ e.jsx("title", { children: b }),
          p && /* @__PURE__ */ e.jsx("meta", { name: "description", content: p })
        ] }),
        /* @__PURE__ */ e.jsxs(C, { className: "max-w-full xl:w-full xl:max-w-3xl flex-1 shrink pt-(--padding-content-top)", children: [
          (c || o) && /* @__PURE__ */ e.jsxs("header", { children: [
            c && /* @__PURE__ */ e.jsx(k, { children: c }),
            o && /* @__PURE__ */ e.jsx(h, { level: 1, id: T(o), children: o })
          ] }),
          /* @__PURE__ */ e.jsx(
            t,
            {
              components: { ...A(), ..._ }
            }
          ),
          /* @__PURE__ */ e.jsx("div", { className: "h-16" }),
          f && r || g ? /* @__PURE__ */ e.jsxs("div", { className: "flex justify-between text-xs text-muted-foreground ", children: [
            /* @__PURE__ */ e.jsx("div", {}),
            /* @__PURE__ */ e.jsxs("div", { className: "flex items-center gap-2", children: [
              /* @__PURE__ */ e.jsx("div", { children: g && /* @__PURE__ */ e.jsx(L, { asChild: !0, variant: "ghost", size: "sm", children: /* @__PURE__ */ e.jsxs(
                "a",
                {
                  href: g,
                  target: "_blank",
                  rel: "noopener noreferrer",
                  className: "flex items-center gap-1",
                  children: [
                    /* @__PURE__ */ e.jsx(P, { size: 12 }),
                    y
                  ]
                }
              ) }) }),
              /* @__PURE__ */ e.jsx("div", { children: f && r && /* @__PURE__ */ e.jsxs(
                "div",
                {
                  title: r.toLocaleString(void 0, {
                    dateStyle: "full",
                    timeStyle: "medium"
                  }),
                  children: [
                    "Last modified on",
                    " ",
                    /* @__PURE__ */ e.jsx("time", { dateTime: r.toISOString(), children: r.toLocaleDateString("en-US", {
                      dateStyle: "long"
                    }) })
                  ]
                }
              ) })
            ] })
          ] }) : null,
          !N && /* @__PURE__ */ e.jsxs(e.Fragment, { children: [
            /* @__PURE__ */ e.jsx("div", { className: "h-px bg-border mt-2 mb-6" }),
            /* @__PURE__ */ e.jsx(
              R,
              {
                prev: m ? { to: m.id, label: m.label ?? "" } : void 0,
                next: x ? { to: x.id, label: x.label ?? "" } : void 0,
                className: "mb-10"
              }
            )
          ] })
        ] }),
        /* @__PURE__ */ e.jsx("div", { className: "hidden xl:block", "data-pagefind-ignore": "all", children: E && /* @__PURE__ */ e.jsx(H, { entries: u }) })
      ]
    }
  );
};
export {
  Y as MdxPage
};
