// zudoku/lib/OasProvider-B6Co-6Vj.js
import { j as s } from "./jsx-runtime-DywqP_6a.js";
import { useMemo as m } from "react";
import { O as l } from "./chunk-PVWAREVJ-BO6B-RAk.js";
import { d as j, O as c, G as O } from "./index-BRQDtecD.js";
const d = (...r) => {
  const e = r.filter((t) => !!t).map((t) => t.replace(/(^\/+|\/+$)/g, "")).join("/").replace(/(^\/+|\/+$)/g, "");
  return e ? `/${e}` : "/";
}, P = ({
  basePath: r,
  version: e,
  config: t,
  client: n
}) => {
  const a = m(() => {
    const p = j(t), i = Object.values(t.input).at(0), u = t.type === "file" ? {
      type: t.type,
      // biome-ignore lint/style/noNonNullAssertion: is guaranteed to be defined
      input: e ? t.input[e] : i
    } : { type: t.type, input: t.input };
    return {
      config: {
        ...t,
        version: e ?? p.at(0),
        versions: Object.fromEntries(
          p.map((o) => [o, d(r, o)])
        ),
        ...u
      }
    };
  }, [t, r, e]);
  return /* @__PURE__ */ s.jsx(c, { value: a, children: /* @__PURE__ */ s.jsx(O, { client: n, children: /* @__PURE__ */ s.jsx(l, {}) }) });
};
export {
  P as OasProvider
};
