// zudoku/lib/zudoku.plugin-api-catalog.js
import { j as t } from "./jsx-runtime-DywqP_6a.js";
import { s as f } from "./index-DhL3LhWV.js";
import { d as b, m as j } from "./chunk-PVWAREVJ-BO6B-RAk.js";
import { b as x, j as d } from "./ZudokuContext-BcqSiygJ.js";
import { H as v } from "./index.esm-BnYHxCYC.js";
import { L as y } from "./index-CVfG8Pb4.js";
import { u as N } from "./hook-BkLXq8pm.js";
import { H as k, M as S } from "./ErrorAlert-xpVkFS_n.js";
const w = ({
  items: n,
  filterCatalogItems: r = (i) => i,
  label: c = "API Library",
  categoryLabel: s
}) => {
  const i = N(), a = b({ path: "/catalog/:category" })?.params.category, o = x({
    queryFn: () => r(n, { auth: i }),
    queryKey: ["catalogItems", i]
  }), m = a ? { "data-pagefind-ignore": "all" } : {};
  return /* @__PURE__ */ t.jsxs("section", { className: "pt-(--padding-content-top)", ...m, children: [
    /* @__PURE__ */ t.jsx(v, { children: /* @__PURE__ */ t.jsxs("title", { children: [
      s ? `${s} - ` : "",
      c
    ] }) }),
    /* @__PURE__ */ t.jsxs("div", { className: "grid gap-4", children: [
      /* @__PURE__ */ t.jsxs(k, { level: 2, children: [
        c,
        s && ` - ${s}`
      ] }),
      /* @__PURE__ */ t.jsx("div", { className: "grid grid-cols-2 gap-4", children: o.data.filter(
        (e) => !a || e.categories.find(
          (l) => l.tags.find((g) => h(l.label, g) === a)
        )
      ).map((e) => /* @__PURE__ */ t.jsx(
        y,
        {
          to: d(e.path),
          className: "no-underline hover:!text-foreground",
          children: /* @__PURE__ */ t.jsxs("div", { className: "border h-full rounded-lg p-4 flex flex-col gap-2 cursor-pointer hover:bg-border/20 font-normal", children: [
            /* @__PURE__ */ t.jsx("span", { className: "font-semibold", children: e.label }),
            /* @__PURE__ */ t.jsx(
              S,
              {
                className: "text-sm whitespace-pre-wrap mb-6 line-clamp-2",
                content: e.description,
                components: {
                  // Because we're wrapping the description in a Link already,
                  // we need to strip out other links to not get a hydration error, like:
                  // > In HTML, <a> cannot be a descendant of <a>.
                  // > This will cause a hydration error.
                  a: (l) => /* @__PURE__ */ t.jsx("span", { ...l })
                }
              }
            )
          ] })
        },
        e.path
      )) })
    ] })
  ] });
}, h = (n, r) => f(`${n}-${r}`), A = ({
  path: n,
  items: r,
  label: c,
  categories: s = [],
  filterCatalogItems: i
}) => {
  const p = Object.fromEntries(
    s.flatMap(
      (a) => [void 0, ...a.tags].map((o) => [
        d(n, o ? h(a.label, o) : void 0),
        o
      ])
    )
  );
  return {
    getNavigation: async (a) => {
      if (!Object.keys(p).some(
        (e) => j(e, a)
      ))
        return [];
      const m = s.map((e) => ({
        type: "category",
        label: e.label,
        collapsible: !1,
        items: e.tags.map((l) => ({
          type: "link",
          to: d(n, h(e.label, l)),
          label: l,
          badge: {
            label: String(
              r.filter(
                (g) => g.categories.find((u) => u.tags.includes(l))
              ).length
            ),
            color: "outline"
          }
        }))
      }));
      return m.unshift({
        type: "link",
        to: d(n),
        label: "Overview",
        badge: { label: String(r.length), color: "outline" }
      }), m;
    },
    getRoutes: () => Object.entries(p).map(([a, o]) => ({
      path: a,
      element: /* @__PURE__ */ t.jsx(
        w,
        {
          label: c,
          categoryLabel: o,
          items: r,
          filterCatalogItems: i,
          categories: s
        }
      )
    }))
  };
};
export {
  A as apiCatalogPlugin,
  h as getKey
};
